src/Pelagos/Util/DOIutil.php

```diff
@@ -375,4 +352,72 @@ function ($matches) {
             $input
         );
     }
+
+    /**
+     * This function generates a new DOI string that is not already in use per datacite.
+     *
+     * @throws HttpClientErrorException When there was an 4xx error negotiating with Datacite, except for a 400 or 404.
+     * @throws HttpServerErrorException When there was an 5xx error negotiating with Datacite.
+     * @throws HttpClientErrorException When there was an 4xx error negotiating with Datacite, except for a 400 or 404, in loop.
+     * @throws HttpServerErrorException When there was an 5xx error negotiating with Datacite, in loop.
+     *
+     * @return string A unique, unused DOI string.
+     */
+    public function generateDoi()
+    {
+        $encoder = new Base32CrockfordEncoder();
+        // 1099511627775 encodes to the longest 8 character Crockford 32 string.
+        $max = 1099511627775;
+        // Start at 1 (0 is problematic as library does not produce checksum for 0.)
+        $random = mt_rand(1, $max);
+        // Add prefix and remove the checksum character.
+        $doi = $this->doiprefix . '/' . substr($encoder->encode($random), 0, -1);
+        try {
+            $doiExists = $this->checkDoiExistsExternal($doi);
+        } catch (HttpClientErrorException $exception) {
+            throw new HttpClientErrorException($exception->getMessage(), $exception->getCode());
+        } catch (HttpServerErrorException $exception) {
+            throw new HttpServerErrorException($exception->getMessage(), $exception->getCode());
+        }
+        while ($doiExists) {
+            $random = mt_rand(1, $max);
```

Comment:
This whole block of code is used twice, it can be created into a separate function

Reply:
yes, but each call would need to be run in a try/catch...